### PR TITLE
feat: Add renaming of files and folders ✌️

### PR DESCRIPTION
**Pull request by kosssi**

### CODE REVIEW

File: src/components/FilenameInput.jsx [hunk 1]
```diff
@@ -49,7 +51,7 @@ export default class FilenameInput extends Component {
 
   submit () {
     this.setState({ working: true, error: false })
-    this.props.onSubmit(this.state.value)
+    this.props.onSubmit && this.props.onSubmit(this.state.value)
```
Comment (enguerran) [inline src/components/FilenameInput.jsx]: We shall use [PropTypes](https://facebook.github.io/react/docs/typechecking-with-proptypes.html#proptypes) and its `isRequired` API instead of this code.

**Comment by kosssi** [comment]
[(1) [FIL-45] Renommer un élément](https://trello.com/c/cgLOntPJ/276-1-fil-45-renommer-un-element)

File: src/components/File.jsx [hunk 2]
```diff
@@ -94,6 +98,20 @@ class File extends Component {
     )
   }
 
+  renderFilenameInput (attributes) {
```
Comment (enguerran) [inline src/components/File.jsx]: I think this shall be a component taking `isRenaming`, `attributes`, `opening`, `selectionModeActive` and returning either a sub component called `RenameInput` or `FileCell`. It worth be clearer and more homogeneous.

Comment (enguerran) [inline src/components/File.jsx]: In other words, there is no reason that the File Component embeds functions that are not related to the File Component as *display a input field* or *display a *rich filename cell*.

File: src/components/FileList.jsx [hunk 3]
```diff
@@ -83,19 +83,22 @@ class FileList extends PureComponent {
...
     } = this.props
     const file = this.props.files[index]
     if (!file) {
       return <FilePlaceholder key={key} style={style} />
     }
+    const { isRenaming, renamingFile } = this.props
+    const isFileRenaming = isRenaming && renamingFile && renamingFile.id === file.id
+    const isSelected = selected.find(f => f && f.id === file.id) !== undefined
     return (
       <File
```
Comment (enguerran) [inline src/components/FileList.jsx]: I know it is not this PR's purpose, but I would like to highlight that this component **smells a lot**: too many attributes makes the component hard to test and it bears too many responsibilities.

File: src/ducks/files/Container.jsx [hunk 4]
```diff
@@ -18,6 +19,8 @@ import {
 const mapStateToProps = (state, ownProps) => ({
   isTrashContext: false,
   canUpload: true,
+  isRenaming: isRenaming(state),
```
Comment (enguerran) [inline src/ducks/files/Container.jsx]: Not sure that `isRenaming` should be hosted in the global state. It may stay in a stateful component responsible of the file name edition.

File: src/ducks/files/Container.jsx [hunk 5]
```diff
@@ -38,6 +41,12 @@ const mapDispatchToProps = (dispatch, ownProps) => ({
         confirm(<DeleteConfirm t={ownProps.t} fileCount={files.length} />)
           .then(() => dispatch(trashFiles(files)))
           .catch(() => {})
+    },
+    singleSelection: {
+      rename: selected => new Promise(resolve => {
+        dispatch(startRenaming(selected[0]))
+        resolve()
+      })
```
Comment (enguerran) [inline src/ducks/files/Container.jsx]: Why is there a `Promise` here? As we use `redux-thunk` we already have async dispatch. It adds complexity.

File: src/ducks/files/RenameInput.jsx [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+const mapStateToProps = (state, ownProps) => ({
+  name: getUpdatedName(state)
+})
+
+const mapDispatchToProps = (dispatch, ownProps) => ({
+  updateRenaming: name => dispatch(updateFileName(name)),
+  rename: () => dispatch(rename()),
+  abortRenaming: name => dispatch(abortRenaming())
```
Comment (enguerran) [inline src/ducks/files/RenameInput.jsx]: Not sure the `renaming` workflow should be hosted in a stateful component, not in the application's global state. Cf. [“You Might Not Need Redux”](https://medium.com/@dan_abramov/you-might-not-need-redux-be46360cf367)

File: src/ducks/files/files.js [hunk 7]
```diff
@@ -0,0 +1,2 @@
+// selectors
+export const getFiles = state => state.view.files
```
Comment (enguerran) [inline src/ducks/files/files.js]: Why do you divide the files duck into several tiny parts? We should think about logical responsibilities to organize files, functions and components.

File: src/ducks/files/rename.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
+/* global cozy */
+
+import { getFiles } from './files'
+import { META_DEFAULTS as meta, extractFileAttributes } from '../../actions'
+
+// constants
+
+const START_RENAMING = 'START_RENAMING'
+const ABORT_RENAMING = 'ABORT_RENAMING'
+export const RENAME_SUCCESS = 'RENAME_SUCCESS'
```
Comment (enguerran) [inline src/ducks/files/rename.js]: Regarding  the ducks architecture, there is no need to export a constant as the ducks should contain everything necessary and sufficient to perform redux's tasks.

File: src/ducks/files/rename.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+      return state
+  }
+}
+export default renameReducer
+
+// selectors
+
+export const isRenaming = state => state.rename !== initialState
+export const getRenamingFile = state => state.rename.file
+export const getUpdatedName = state => state.rename.name
```
Comment (enguerran) [inline src/ducks/files/rename.js]: The selectors of a sub redux state should correspond to that specific sub redux state as mentioned at 1'05 in the [Redux: Colocating Selectors with Reducers](https://egghead.io/lessons/javascript-redux-colocating-selectors-with-reducers) egghead video.

File: src/ducks/files/rename.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+// action creators async
+
+export const rename = () => async (dispatch, getState) => {
+  const state = getState()
+  const files = getFiles(state)
+  const renamingFile = getRenamingFile(state)
+  const updatedName = getUpdatedName(state)
+  const isExisting = files.find(f => f.name === updatedName)
+
+  if (isExisting) {
```
Comment (enguerran) [inline src/ducks/files/rename.js]: Rename action should be a step-by-step workflow, not all-in-one big action. I mean, it will be clearer to have an cleaner API:

```js
dispatch(checkFilename(files, newFilename))
    .then(filename => dispatch(rename(id, filename)))
    .catch(err => Throw new AlreadyExistingFilenameError)
```

IMHO

File: src/reducers/view.js [hunk 11]
```diff
@@ -100,6 +105,8 @@ const files = (state = [], action) => {
       const clone = state.slice(0)
       action.files.forEach((f, i) => { clone[action.skip + i] = f })
       return clone
+    case RENAME_SUCCESS:
+      return updateItem(action.file, state)
```
Comment (enguerran) [inline src/reducers/view.js]: Not sure I understand what's happening here:
> In that reducer, if an action comes and is called `RENAME_SUCCESS`, then we return a new state given by a function called `updateItem`. Why does it stand in a `view.js` reducer and not in something like a `file` reducer?

Comment (kosssi) [inline src/reducers/view.js]: `view.js` is very big ;) it's reducers of files. We will have to refactoring a lot.

Comment (kosssi) [inline src/reducers/view.js]: Vu que l'on change le nom du fichier, il faut mettre à jour l'ordre de la liste des fichiers/dossiers

File: src/ducks/files/Container.jsx (hunk 5 above)
Comment (kosssi) [inline src/ducks/files/Container.jsx]: c'est un mini hack... https://github.com/cozy/cozy-drive/blob/master/src/components/FileActionMenu.jsx#L24 il faut que onClick retourne une promise.

Comment (enguerran) [inline src/ducks/files/Container.jsx]: Behhh, du coup, pourquoi c'est pas [`startRenaming`](https://github.com/cozy/cozy-drive/pull/249/files#diff-860b84dc6e5ad9b816a7d46dcab6c95dR41) qui retourne une promesse avec du redux-thunk ou bien encore `onClick` directement puisque le besoin est UI ?

File: src/ducks/files/rename.js (hunk 8 above)
Comment (kosssi) [inline src/ducks/files/rename.js]: Je comprends, mais en même temps je vois pas comment on peut réordonner la liste des fichiers sans lui dire explicitement : https://github.com/cozy/cozy-drive/pull/249/files#diff-b228a6ff1c0c9483294c6306d787236aR108

File: src/ducks/files/files.js (hunk 7 above)
Comment (kosssi) [inline src/ducks/files/files.js]: a la base je l'avais mis dans index.js, mais vu qu'il y avait une dependence circulaire je me suis dit que il vallait mieux le séparer du container qui est importé dans index.js ;)

File: src/reducers/view.js (hunk 11 above)
Comment (enguerran) [inline src/reducers/view.js]: Dans l'absolu, pour afficher une liste ordonnée avec react et redux, on va créer une state avec la liste, un selector avec la liste ordonnée et un component qui `mapStateToProps` avec un `shouldComponentUpdate` si on a des besoins de perfs. Exemple : https://codesandbox.io/s/0R6rrOJB7
